README.rst

Mbmore!
==============
This repository is a collection of custom Cyclus facility archetypes that
utilize a random number generator (RNG) to create non-deterministic behaviors.
General methods controlling the behavior (including random number generation
and Gaussian distributions) are defined in the `behavior functions. <https://github.com/cnerg/mbmore/blob/master/src/behavior_functions.h>`_


Behavior Functions
------------------
These functions use the C++ 'srand' to create behaviors that change in time.
The RNG is seeded only once per simulation.  The seed value can be controlled
by the <rng_seed> tag in individual archetypes. (Although there are rng_seed
inputs for each archetype, it is only set once, so avoid defining it multiple
times in one input file). If set to -1, rng_seed is seeded on the system time at
simulation execution. Otherwise RNG is seeded on the value of rng_seed, for
reproducibility.

Available behavior functions are:

* *CalcYVal*: For various functions, returns Y given a set of relevant constants and X.
  
  - ``Constant`` [y0]:  a constant line with the form y = y0
  - ``Linear`` [y0, m]: line with the form y = y0 + m*x
  - ``Power`` [A, (B=1)]: powerlaw fn of the form y = B*(x^A). B defaults to 1 if not specified.
  - ``Step`` [y0, yf, (T)]: step function of the form y = y0  (x < T), y = y1 (x >= T).

* *EveryXTimestep* - Returns true every X interval
* *EveryRandomXTimestep* - Returns true with an approximate frequency defined by X, with individual instances randomly determined.
* *RNG_Integer* - Returns a randomnly choses discrete number between the defined min and max.
* *RNG_NormalDist* - Returns a randomnly generated number from a normal distribution defined by a mean and a sigma (full-width-half-max)
* *XLikely* - Returns true with an average likelihood defined by X [0-1], with individual instances randomly determined. 



Archetypes
----------

CascadeEnrich
+++++++++++++
Based on `cycamore:Enrich <http://fuelcycle.org/user/cycamoreagents.html#cycamore-enrichment>`_ , this facility designs a cascade based on the physical parameters of the individual counter-current centrifuges being used, the target assays and cascade feed flow, and the available number of centrifuges. The cascade is designed as an ideal one-up, one-down cascade in which the product/tails from one stage moves up/down to the next stage, respectively. Centrifuge machine performance is calculated using the Ratz equation and assuming an R2 (U-238) withdrawl radius of 0.975*a (radius of the centrifuge).  Only one physical centrifuge design maybe used in the cascade, it cannot mix centrifuge designs.  The cascade produced deviates from an ideal cascade only in that integer numbers of centrifuges must be used for each stage.  If the number of available centrifuges is insufficient to meet both the target assays and the target feed flow, then feedflow will be reduced to meet the other requirements.  The cascade will produce an enrichment level At Least as high as the ``design_product_assay``, again constrained by integer stage steps.   This archetype automatically designs the cascade at the beginning of the simulation, at which point the physical configuration of the centrifuges and cascade design are fixed for the remainder of the simulation.  The facility will still attempt to produce material of non-target product assay as requested, within the limits of integer number of stages, SWU and feed flow capacity constraints. When processing off-design material assays, separative capacity will be reduced. The facility assumes two-isotope enrichment (U-235 and U-238) such that molecutlar mass is 0.352kg/mol (UF6), a cut (ratio of product/feed quantity) of 0.5, an internal flow of 2.0 (in practice dependent on baffle/scoop design and can range from 2-4), and a pressure ratio of 1000 (Glaser, Science and Global Security, 2009).

Future work: Cut, efficiency (which can be significantly less than 1), pressure ratio, and internal flow should be user-defined with reasonable defaults. Blending capability to achieve the exact requested enrichment level. R2 withdrawl radius should be user defined as well (called in enrich_functions::CalcDelU). Time-based calculations (flow rates, SWU etc) should be changed to use arbitrary time base, currently timesteps of one month are assumed.

Enrichment and cascade design calculations are implemented in the accompanying enrich_functions.cc file.
  - ``design_feed_flow``: The amount of feed material the cascade is
    initially designed to process (kg/month).  Combined with
    ``design_feed_assay``, ``design_product_assay``, and ``design_waste_assay``,
    this sets the ideal
    number of centrifuges for the cascade. If there are not sufficient
    centrifuges available then actual production will not achieve this target.
  - ``max_centrifuges``: The total number of centrifuges available to design
    the cascade. If there are more available than necessary to achieve the
    ``design_feed_flow``, (given ``design_feed_assay``, ``design_product_assay``
    and ``design_waste_assay``)  then the cascade will use as many of the
    centrifuges as possible given the constraint that each stage must have an
    integer number of machines.
  - ``design_feed_assay``: Expected feed assay for which ideal cascade
    is designed.
  - ``design_product_assay``: Desired product assay for which ideal cascade
    is designed.
  - ``design_tails_assay``: Desired waste assay for which ideal cascade
    is designed.
  - ``temp``: Operating temperature of individual centrifuge machine, which
    is used to calculate machine separative capacity (Kelvin).
  -``centrifuge_velocity``: Desired operational velocity of the centrifuge,
    which is used to calculate machine separative capacity (m/s).
  - ``height``: height of the centrifuge, which is used to calculate machine
    separative capacity (m).
  - ``diameter``: diameter of the centrifuge, which is used to calculate machine
    separative capacity (m).
  - ``machine_feed``: maximum throughput for a single centrifuge, which is used
    to calculate machine separative capacity (m).

    



InteractRegion
++++++++++++++
This manager region is used to study the likelihood of a state pursuing and acquiring a nuclear weapon given relationships with a set of neighboring states, represented by StateInst.  The InteractRegion is a super-region that includes All States in the simulation and acts as a 'Simulation Context' for universal information such as the functional form and weighting of the Pursuit and Acquire decision-making equations. The Pursuit and Acquire Equations define the level of motivation for the country on a scale of 0-10, The Likely Equation (LikelyEqn) rescales this motivation into a likelihood of action at each timestep (0-1).
 - ``acquire_weights`` (not implemented): A map of (Factor Name, Weight). The equation that defines how motivated a state is to acquire a weapon on a given timestep; these are the relative weights of each of the factors considered in the determinination. The sum of all of the weights should be 1, and any factor whose weight is not defined here will be ignored in the calculation (weight = 0). Factor names are case sensitive and should match those defined in the StateInst.
 - ``pursuit_weights``: A map of (Factor Name, Weight). The equation that defines how motivated a state is pursue a weapon on a given timestep; these are the relative weights of each of the factors considered in the determinination. The sum of all of the weights should be 1, and any factor whose weight is not defined here will be ignored in the calculation (weight=0). Factor names are case sensitive and should match those defined in the StateInst.
 - ``likely_convert``: A map of (Equation, (Form, <Constants>)). After the motivational equation (Pursuit or Acquire) returns a value between 0 and 10, ``likely_convert`` rescales that into a likelihood of taking action (0-1).  The form of the equation can be defined to be any of those available in the behavior_function method *CalcYVal*.  For example, ('Pursuit', ('Power', [2])) means LikelyEqn = (Pursuit/10)^2. Then for Pursuit=2, LikelyEqn = 0.04, while for Pursuit = 9, LikelyEqn = 0.81.  The StateInst uses the result of LikelyEqn to convert to Yes or No decision for the timestep.
 - ``p_conflict_map``: A map of (Primary State, (Secondary State, Relation)) that defines the conflict between each pair of states at t=0.  Each state pairing must be defined (i.e. separate entries for StateA-StateB and StateB-StateA).  Options are +1 (friendly), 0 (neutral), -1 (antagonistic).  If states are in agreement about their mutual relationships, ``symmetric`` should be set to 1 (True). Otherwise states can have inconsistent perceptions of one another. Dynamic changes to  the conflict between two states are applied using the StateInst ``pursuit_factors`` variable.
 - ``symmetric`` (default 0): If 1 (True) then any changes in conflict between two states (StateA-StateB) will be mirrored also (StateB-StateA will have the same value). Otherwise if set to 0 (False) then states can have mutually inconsistent perceptions.  This flag affects only Changes to the relationships (defined StateInst), does not force initial conflict values to be symmetric.

A note on *Conflict*. Conflict is an interactive factor between states in the simulation. It is defined by a combination of relationship between states (enemy, ally or neutral) as well as the weapons status of each state. It updates in time as weapons status changes.  Each state-pair receives a conflict score between 0-10 based on `this table. <https://docs.google.com/document/d/1c9YeFngXm3RCbuyFCEDWJjUK9Ovn072SpmlZU6j1qhg/edit?usp=sharing>`_ . In a simulation with more than 2 states, the net conflict score for state A is the average of its individual pair conflict scores with B, C, D.. . .

StateInst
+++++++++
This manager institution is used along with InteractRegion to study whether a state will pursue or acquire a nuclear weapon given a set of political or economic internal Factors, as well as its relationships with a set of neighboring states.  At each timestep, the state decides whether or not to pursue a nuclear weapon by calculating the Pursuit Equation using these Factors (the relative weights of the factors are defined in the InteractRegion).  If the state decides to Pursue, then on the next timestep, a Secret Enrichment Facility and a Secret Receiver (sink) are deployed. The pursuit equation continues to be calculated at each timestep, and its value is used to determine whether the stae has succeeded in acquiring a weapon. If the state succeeds in Acquiring at time T, then HEU is produced at (T+1), and it is moved to the Receiver at (T+2), the quantity of HEU produced is defined in the input file as the requested quantity for the secret sink.
  - ``acquire_factors``: Not supported (see ``pursuit_factors`` for reference)

changelog.rst
PR #17
------

**Added:** None 

**Changed:** 
 - moved the SortsBids method in the behavior_functions as it is needed in
   RandomEnrich archetypes.

**Deprecated:** None

**Removed:** 
 - removed CascadeEnrich initialy developed by mbmcgarry (not working).

**Fixed:** None

**Security:** None


PR #19
------

**Added:** 
  - add centrifuges definitions, aims to be used in the future CascadeEnrich
    archetype. 

**Changed:** None 

**Deprecated:** None

**Removed:** None

**Fixed:** None

**Security:** None




src/RandomEnrich.cc
// Implements the RandomEnrich class
#include "RandomEnrich.h"
#include "behavior_functions.h"
#include "sim_init.h"

#include <algorithm>
#include <boost/lexical_cast.hpp>
#include <cmath>
#include <limits>
#include <sstream>
#include <vector>

namespace mbmore {


// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RandomEnrich::RandomEnrich(cyclus::Context* ctx)
    : cyclus::Facility(ctx),
      tails_assay(0),
      sigma_tails(0),
      social_behav("None"),
      behav_interval(0),
      heu_ship_qty(0),
      inspect_freq(0),
      n_swipes(10),
      false_pos(0),
      false_neg(0),
      rng_seed(0),
      swu_capacity(0),
      max_enrich(1),
      initial_feed(0),
      feed_commod(""),
      feed_recipe(""),
      product_commod(""),
      tails_commod(""),
      order_prefs(true) {}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RandomEnrich::~RandomEnrich() {}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
std::string RandomEnrich::str() {
  std::stringstream ss;
  ss << cyclus::Facility::str() << " with enrichment facility parameters:"
     << " * SWU capacity: " << SwuCapacity()
     << " * Tails assay: " << curr_tails_assay
     << " * Feed assay: " << FeedAssay()
     << " * Input cyclus::Commodity: " << feed_commod
     << " * Output cyclus::Commodity: " << product_commod
     << " * Tails cyclus::Commodity: " << tails_commod;
  return ss.str();
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::Build(cyclus::Agent* parent) {
  using cyclus::Material;

  Facility::Build(parent);
  if (initial_feed > 0) {
    inventory.Push(Material::Create(this, initial_feed,
                                    context()->GetRecipe(feed_recipe)));
  }

  LOG(cyclus::LEV_DEBUG2, "EnrFac") << "RandomEnrich "
                                    << " entering the simuluation: ";
  LOG(cyclus::LEV_DEBUG2, "EnrFac") << str();
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::Tick() {
  int cur_time = context()->time();

  // set inspection defaults
  if (cur_time == 0) {
    net_heu = 0;
    HEU_present = 0;
  }

  // decide whether trading if trading only sometimes.
  trade_timestep = 0;
  if (social_behav == "Every" && behav_interval > 0) {
    trade_timestep = (EveryXTimestep(cur_time, behav_interval));
  } else if (social_behav == "Random" && behav_interval > 0) {
    trade_timestep = (EveryRandomXTimestep(behav_interval, rng_seed));
  } else if (social_behav == "None") {
    trade_timestep = 1;
  }

  // determine tails assay for the timestep if it is variable
  curr_tails_assay = RNG_NormalDist(tails_assay, sigma_tails, rng_seed);
  if (curr_tails_assay < (tails_assay - sigma_tails)) {
    curr_tails_assay = tails_assay - sigma_tails;
  }
  if (curr_tails_assay > (tails_assay + sigma_tails)) {
    curr_tails_assay = tails_assay + sigma_tails;
  }

  LOG(cyclus::LEV_INFO3, "EnrFac") << prototype() << " is ticking {";
  LOG(cyclus::LEV_INFO3, "EnrFac") << "}";
  current_swu_capacity = SwuCapacity();
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::Tock() {
  using cyclus::toolkit::RecordTimeSeries;
  RecordTimeSeries<cyclus::toolkit::ENRICH_SWU>(this, intra_timestep_swu_);
  RecordTimeSeries<cyclus::toolkit::ENRICH_FEED>(this, intra_timestep_feed_);

  // Add any inspections to the Inspection table
  bool do_inspect = EveryRandomXTimestep(inspect_freq, rng_seed);
  if (do_inspect == true) {
    RecordInspection_();
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
std::set<cyclus::RequestPortfolio<cyclus::Material>::Ptr>
RandomEnrich::GetMatlRequests() {
  using cyclus::Material;
  using cyclus::RequestPortfolio;
  using cyclus::Request;

  std::set<RequestPortfolio<Material>::Ptr> ports;
  RequestPortfolio<Material>::Ptr port(new RequestPortfolio<Material>());
  Material::Ptr mat = Request_();
  double amt = mat->quantity();

  if (amt > cyclus::eps()) {
    port->AddRequest(mat, this, feed_commod);
    ports.insert(port);
  }

  return ports;
}
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// Sort offers of input material to have higher preference for more
//  U-235 content
void RandomEnrich::AdjustMatlPrefs(
    cyclus::PrefMap<cyclus::Material>::type& prefs) {
  using cyclus::Bid;
  using cyclus::Material;
  using cyclus::Request;

  if (order_prefs == false) {
    return;
  }

  cyclus::PrefMap<cyclus::Material>::type::iterator reqit;

  // Loop over all requests
  for (reqit = prefs.begin(); reqit != prefs.end(); ++reqit) {
    std::vector<Bid<Material>*> bids_vector;
    std::map<Bid<Material>*, double>::iterator mit;
    for (mit = reqit->second.begin(); mit != reqit->second.end(); ++mit) {
      Bid<Material>* bid = mit->first;
      bids_vector.push_back(bid);
    }
    std::sort(bids_vector.begin(), bids_vector.end(), SortBids);

    // Assign preferences to the sorted vector
    double n_bids = bids_vector.size();
    bool u235_mass = 0;

    for (int bidit = 0; bidit < bids_vector.size(); bidit++) {
      int new_pref = bidit + 1;

      // For any bids with U-235 qty=0, set pref to zero.
      if (!u235_mass) {
        cyclus::Material::Ptr mat = bids_vector[bidit]->offer();
        cyclus::toolkit::MatQuery mq(mat);
        if (mq.mass(922350000) == 0) {
          new_pref = -1;
        } else {
          u235_mass = true;
        }
      }
      (reqit->second)[bids_vector[bidit]] = new_pref;
    }  // each bid
  }    // each Material Request
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::AcceptMatlTrades(
    const std::vector<std::pair<cyclus::Trade<cyclus::Material>,
                                cyclus::Material::Ptr> >& responses) {
  // see
  // http://stackoverflow.com/questions/5181183/boostshared-ptr-and-inheritance
  std::vector<std::pair<cyclus::Trade<cyclus::Material>,
                        cyclus::Material::Ptr> >::const_iterator it;
  for (it = responses.begin(); it != responses.end(); ++it) {
    AddMat_(it->second);
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
std::set<cyclus::BidPortfolio<cyclus::Material>::Ptr> RandomEnrich::GetMatlBids(
    cyclus::CommodMap<cyclus::Material>::type& out_requests) {
  using cyclus::Bid;
  using cyclus::BidPortfolio;
  using cyclus::CapacityConstraint;
  using cyclus::Converter;
  using cyclus::Material;
  using cyclus::Request;
  using cyclus::toolkit::MatVec;

  std::set<BidPortfolio<Material>::Ptr> ports;

  if ((out_requests.count(tails_commod) > 0) && (tails.quantity() > 0)) {
    BidPortfolio<Material>::Ptr tails_port(new BidPortfolio<Material>());

    std::vector<Request<Material>*>& tails_requests =
        out_requests[tails_commod];
    std::vector<Request<Material>*>::iterator it;
    for (it = tails_requests.begin(); it != tails_requests.end(); ++it) {
      // offer bids for all tails material, keeping discrete quantities
      // to preserve possible variation in composition
      MatVec mats = tails.PopN(tails.count());
      tails.Push(mats);
      for (int k = 0; k < mats.size(); k++) {
        Material::Ptr m = mats[k];
        Request<Material>* req = *it;
        tails_port->AddBid(req, m, this);
      }
    }
    // overbidding (bidding on every offer)
    // add an overall capacity constraint
    CapacityConstraint<Material> tails_constraint(tails.quantity());
    tails_port->AddConstraint(tails_constraint);
    LOG(cyclus::LEV_INFO5, "EnrFac") << prototype()
                                     << " adding tails capacity constraint of "
                                     << tails.capacity();
    ports.insert(tails_port);
  }

  if ((out_requests.count(product_commod) > 0) && (inventory.quantity() > 0)) {
    // BidPortfolio<Material>::Ptr commod_port(new BidPortfolio<Material>());
    BidPortfolio<Material>::Ptr commod_port =
        ConsiderMatlRequests(out_requests);

    /*
    std::vector<Request<Material>*>& commod_requests =
      out_requests[product_commod];
    std::vector<Request<Material>*>::iterator it;
    for (it = commod_requests.begin(); it != commod_requests.end(); ++it) {
      Request<Material>* req = *it;
      Material::Ptr mat = req->target();
      double request_enrich = cyclus::toolkit::UraniumAssay(mat) ;

      if (ValidReq(req->target()) && (request_enrich <= max_enrich)) {
        Material::Ptr offer = Offer_(req->target());
        commod_port->AddBid(req, offer, this);
      }
    }
    */
    Converter<Material>::Ptr sc(
        new SWUConverter(FeedAssay(), curr_tails_assay));
    Converter<Material>::Ptr nc(
        new NatUConverter(FeedAssay(), curr_tails_assay));
    CapacityConstraint<Material> swu(swu_capacity, sc);
    CapacityConstraint<Material> natu(inventory.quantity(), nc);
    commod_port->AddConstraint(swu);
    commod_port->AddConstraint(natu);

    LOG(cyclus::LEV_INFO5, "EnrFac")
        << prototype() << " adding a swu constraint of " << swu.capacity();
    LOG(cyclus::LEV_INFO5, "EnrFac")
        << prototype() << " adding a natu constraint of " << natu.capacity();
    ports.insert(commod_port);
  }
  return ports;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
bool RandomEnrich::ValidReq(const cyclus::Material::Ptr mat) {
  cyclus::toolkit::MatQuery q(mat);
  double u235 = q.atom_frac(922350000);
  double u238 = q.atom_frac(922380000);
  return (u238 > 0 && u235 / (u235 + u238) > curr_tails_assay);
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::GetMatlTrades(
    const std::vector<cyclus::Trade<cyclus::Material> >& trades,
    std::vector<std::pair<cyclus::Trade<cyclus::Material>,
                          cyclus::Material::Ptr> >& responses) {
  using cyclus::Material;
  using cyclus::Trade;

  intra_timestep_swu_ = 0;
  intra_timestep_feed_ = 0;

  std::vector<Trade<Material> >::const_iterator it;
  for (it = trades.begin(); it != trades.end(); ++it) {
    double qty = it->amt;
    std::string commod_type = it->bid->request()->commodity();
    Material::Ptr response;

    // Figure out whether material is tails or enriched,
    // if tails then make transfer of material
    if (commod_type == tails_commod) {
      LOG(cyclus::LEV_INFO5, "EnrFac")
          << prototype() << " just received an order"
          << " for " << it->amt << " of " << tails_commod;
      response = tails.Pop(qty);
    } else {
      LOG(cyclus::LEV_INFO5, "EnrFac")
          << prototype() << " just received an order"
          << " for " << it->amt << " of " << product_commod;
      response = Enrich_(it->bid->offer(), qty);
    }
    responses.push_back(std::make_pair(*it, response));
  }

  if (cyclus::IsNegative(tails.quantity())) {
    std::stringstream ss;
    ss << "is being asked to provide more than its current inventory.";
    throw cyclus::ValueError(Agent::InformErrorMsg(ss.str()));
  }
  if (cyclus::IsNegative(current_swu_capacity)) {
    throw cyclus::ValueError("EnrFac " + prototype() +
                             " is being asked to provide more than" +
                             " its SWU capacity.");
  }
}
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::AddMat_(cyclus::Material::Ptr mat) {
  // Elements and isotopes other than U-235, U-238 are sent directly to tails
  cyclus::CompMap cm = mat->comp()->atom();
  bool extra_u = false;
  bool other_elem = false;
  for (cyclus::CompMap::const_iterator it = cm.begin(); it != cm.end(); ++it) {
    if (pyne::nucname::znum(it->first) == 92) {
      if (pyne::nucname::anum(it->first) != 235 &&
          pyne::nucname::anum(it->first) != 238 && it->second > 0) {
        extra_u = true;
      }
    } else if (it->second > 0) {
      other_elem = true;
    }
  }
  if (extra_u) {
    cyclus::Warn<cyclus::VALUE_WARNING>(
        "More than 2 isotopes of U.  "
        "Istopes other than U-235, U-238 are sent directly to tails.");
  }
  if (other_elem) {
    cyclus::Warn<cyclus::VALUE_WARNING>(
        "Non-uranium elements are "
        "sent directly to tails.");
  }

  LOG(cyclus::LEV_INFO5, "EnrFac") << prototype() << " is initially holding "
                                   << inventory.quantity() << " total.";

  try {
    inventory.Push(mat);
  } catch (cyclus::Error& e) {
    e.msg(Agent::InformErrorMsg(e.msg()));
    throw e;
  }

  LOG(cyclus::LEV_INFO5, "EnrFac")
      << prototype() << " added " << mat->quantity() << " of " << feed_commod
      << " to its inventory, which is holding " << inventory.quantity()
      << " total.";
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
cyclus::Material::Ptr RandomEnrich::Request_() {
  double qty = std::max(0.0, inventory.capacity() - inventory.quantity());
  return cyclus::Material::CreateUntracked(qty,
                                           context()->GetRecipe(feed_recipe));
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
cyclus::Material::Ptr RandomEnrich::Offer_(cyclus::Material::Ptr mat) {
  cyclus::toolkit::MatQuery q(mat);
  cyclus::CompMap comp;
  comp[922350000] = q.atom_frac(922350000);
  comp[922380000] = q.atom_frac(922380000);
  return cyclus::Material::CreateUntracked(
      mat->quantity(), cyclus::Composition::CreateFromAtom(comp));
}
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
cyclus::Material::Ptr RandomEnrich::Enrich_(cyclus::Material::Ptr mat,
                                            double qty) {
  using cyclus::Material;
  using cyclus::ResCast;
  using cyclus::toolkit::Assays;
  using cyclus::toolkit::UraniumAssay;
  using cyclus::toolkit::SwuRequired;
  using cyclus::toolkit::FeedQty;
  using cyclus::toolkit::TailsQty;

  // get enrichment parameters
  double u_assay = UraniumAssay(mat);
  Assays assays(FeedAssay(), u_assay, curr_tails_assay);
  double swu_req = SwuRequired(qty, assays);
  double natu_req = FeedQty(qty, assays);

  // Determine the composition of the natural uranium
  // (ie. U-235+U-238/TotalMass)
  Material::Ptr natu_matl = inventory.Pop(inventory.quantity());
  inventory.Push(natu_matl);

  cyclus::toolkit::MatQuery mq(natu_matl);
  std::set<cyclus::Nuc> nucs;
  nucs.insert(922350000);
  nucs.insert(922380000);
  double natu_frac = mq.mass_frac(nucs);
  double feed_req = natu_req / natu_frac;

  // pop amount from inventory and blob it into one material
  Material::Ptr r;
  try {
    // required so popping doesn't take out too much
    if (cyclus::AlmostEq(feed_req, inventory.quantity())) {
      r = cyclus::toolkit::Squash(inventory.PopN(inventory.count()));
    } else {
      r = inventory.Pop(feed_req);
    }
  } catch (cyclus::Error& e) {
    NatUConverter nc(FeedAssay(), curr_tails_assay);
    std::stringstream ss;
    ss << " tried to remove " << feed_req << " from its inventory of size "
       << inventory.quantity()
       << " and the conversion of the material into natu is "
       << nc.convert(mat);
    throw cyclus::ValueError(Agent::InformErrorMsg(ss.str()));
  }

  // "enrich" it, but pull out the composition and quantity we require from the
  // blob
  cyclus::Composition::Ptr comp = mat->comp();
  Material::Ptr response = r->ExtractComp(qty, comp);
  tails.Push(r);

  current_swu_capacity -= swu_req;

  intra_timestep_swu_ += swu_req;
  intra_timestep_feed_ += feed_req;
  RecordRandomEnrich_(feed_req, swu_req);

  // If enriched to HEU then record total HEU produced
  double heu_definition = 0.2;
  if (u_assay > heu_definition) {
    net_heu += qty;
  }

  LOG(cyclus::LEV_INFO5, "EnrFac") << prototype()
                                   << " has performed an enrichment: ";
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Feed Qty: " << feed_req;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Feed Assay: "
                                   << assays.Feed() * 100;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Product Qty: " << qty;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Product Assay: "
                                   << assays.Product() * 100;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Tails Qty: "
                                   << TailsQty(qty, assays);
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Tails Assay: "
                                   << assays.Tails() * 100;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * SWU: " << swu_req;
  LOG(cyclus::LEV_INFO5, "EnrFac") << "   * Current SWU capacity: "
                                   << current_swu_capacity;

  return response;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::RecordRandomEnrich_(double natural_u, double swu) {
  using cyclus::Context;
  using cyclus::Agent;

  LOG(cyclus::LEV_DEBUG1, "EnrFac") << prototype()
                                    << " has enriched a material:";
  LOG(cyclus::LEV_DEBUG1, "EnrFac") << "  * Amount: " << natural_u;
  LOG(cyclus::LEV_DEBUG1, "EnrFac") << "  *    SWU: " << swu;

  Context* ctx = Agent::context();
  ctx->NewDatum("RandomEnrichs")
      ->AddVal("ID", id())
      ->AddVal("Time", ctx->time())
      ->AddVal("Natural_Uranium", natural_u)
      ->AddVal("SWU", swu)
      ->Record();
}
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void RandomEnrich::RecordInspection_() {
  using cyclus::Context;
  using cyclus::Agent;

  std::string sample_location = "Cascade";

  // TODO: Add multiple samples to an inspection, rules about only Cascade
  // having true positives? Or increased chance of true based on location?
  // TODO: Make HEU definition a State Var (in Tock)

  // If HEU has been made, then we see if a perfect swipe test would find it
  // by calculating HEU_present (ie. maybe heu has been made but swipe is in
  // the wrong spot).  Its presence becomes more likely to detect
  // scaling with time elapsed because we presume there is increasing
  // contamination. Once contamination would be theoretically measured, this
  // contamination remains for the rest of the simulation.
  // (TODO: scale with quantity HEU produced)
  double cur_time = double(context()->time());
  if ((social_behav == "None") && (HEU_present == false)) {
    // HEU is produced continuously (as requested), and removed when some
    // quantity has been
    // produced. Risk of leakage increases with time in discrete steps
    // TODO: This will fail if a facility is producing HEU but never shipping
    // it and inspections are still supposed to occur because it assumes
    // that HEU can only be detected if it has been removed from cascades for
    // shipping.
    std::cout << "Inspect Time: " << cur_time << "  Net HEU produced "
              << net_heu << std::endl;
    if ((net_heu >= heu_ship_qty) && (heu_ship_qty > 0.0)) {
      HEU_present = XLikely(cur_time / (double(simdur) - 1.0), rng_seed);
      std::cout << "HEU Presence? " << HEU_present << std::endl;
      net_heu -= heu_ship_qty;
    }
  } else if ((net_heu > 0.0) && (HEU_present == false)) {
    // HEU is made/shipped at specific intervals defined by behavior fns,
    // so test whether any has been made/shipped since last inspection
    HEU_present = XLikely(cur_time / (double(simdur) - 1.0), rng_seed);
  }

  // Each sample is N swipes, analyzed independently (with a high rate of
  // false readings in practice).
  // Based on whether HEU is 'detected' in the sample, determine whether or not
  // any false positives or negatives change the swipe result.
  int pos_swipes = 0;
  int n_false_pos = 0;
  int n_false_neg = 0;

  for (int swipeit = 0; swipeit < n_swipes; swipeit++) {
    double prob;
    if (HEU_present == true) {
      prob = false_neg;
    } else {
      prob = false_pos;
    }
    bool flip = XLikely(prob, rng_seed);
    //    std::cout << "Flip? " << flip << std::endl;

    // record false positives, false negatives and net 'positive' swipe results
    if (flip) {
      if (!HEU_present) {
        pos_swipes++;
        n_false_pos++;
      } else {
        n_false_neg++;
      }
    } else {
      if (HEU_present) {
        pos_swipes++;
      }
    }

    //    if ((HEU_present && !flip) || (!HEU_present && flip)){
    //      pos_swipes++;
    //    }
  }

  Context* ctx = Agent::context();
  context()
      ->NewDatum("Inspections")
      ->AddVal("AgentID", id())
      ->AddVal("Time", context()->time())
      ->AddVal("SampleLoc", sample_location)
      ->AddVal("FalsePos", double(n_false_pos) / double(n_swipes))
      ->AddVal("FalseNeg", double(n_false_neg) / double(n_swipes))
      ->AddVal("PosSwipeFrac", double(pos_swipes) / double(n_swipes))
      ->Record();

  /*
  LOG(cyclus::LEV_DEBUG1, "EnrFac") << prototype()
                                    << " Has been inspected:";
  LOG(cyclus::LEV_DEBUG1, "EnrFac") << "  * N_Samples: " << natural_u;
  LOG(cyclus::LEV_DEBUG1, "EnrFac") << "  *    Swipe_frac: " << swu;
  */
}
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// Decide whether each individual bid will be responded to based on whether
// the enrichment facility is trading
cyclus::BidPortfolio<cyclus::Material>::Ptr RandomEnrich::ConsiderMatlRequests(
    cyclus::CommodMap<cyclus::Material>::type& out_requests) {
  using cyclus::Bid;
  using cyclus::BidPortfolio;
  using cyclus::Material;
  using cyclus::Request;

  BidPortfolio<Material>::Ptr commod_port(new BidPortfolio<Material>());

  std::vector<Request<Material>*>& commod_requests =
      out_requests[product_commod];
  std::vector<Request<Material>*>::iterator it;
  for (it = commod_requests.begin(); it != commod_requests.end(); ++it) {
    Request<Material>* req = *it;
    Material::Ptr mat = req->target();
    double request_enrich = cyclus::toolkit::UraniumAssay(mat);
    int cur_time = context()->time();
    // if social behavior on and logic says no trade
    if (!trade_timestep) {
      return commod_port;
    }
    if (ValidReq(req->target()) && request_enrich <= max_enrich) {
      Material::Ptr offer = Offer_(req->target());
      commod_port->AddBid(req, offer, this);
    }
  }  // for each out commod

  return commod_port;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
double RandomEnrich::FeedAssay() {
  using cyclus::Material;

  if (inventory.empty()) {
    return 0;
  }
  cyclus::Material::Ptr fission_matl = inventory.Pop(inventory.quantity());
  inventory.Push(fission_matl);
  return cyclus::toolkit::UraniumAssay(fission_matl);
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
extern "C" cyclus::Agent* ConstructRandomEnrich(cyclus::Context* ctx) {
  return new RandomEnrich(ctx);
}

}  // namespace mbmore
